### PR TITLE
Add a formal grammar using ANTLR

### PR DESCRIPTION
**Pull request by sjoelund**
This uses a combined ANTLR3 and ANTLR4 grammar (the differences are
converted from ANTLR3 to ANTLR4). ANTLR4 is used to run the grammar
easily, and ANTLR3 to analyze the grammar to verify it is LL(k=2).

- [ ] `extends_clause` was modified from the 3.4 grammar (no rule annotation exists)
- [ ] `array_arguments`: Changed to use iteration
- [x] `function_call_args` etc were rewritten due to #2250 ~~and also resolves
      ambiguity with named arg or expression
      f(a, b=3) ANTLR4 said "f(a, b" parsed, expecting ")"~~
- [ ] `der_class_specifier` was merged into `short_class_specifier` to get
      k=2 instead of k=3.

Some things to resolve:

- [ ] `equation` resolved the ambiguity by allowing simple expressions
      stand-alone. This is easy to implement for all tools, might be OK for
      flat Modelica, but is usually a user error... How to resolve it? Use
      ANTLR backtracking or predicates? Create a simple_expression_no_cref?
- [ ] There is no script to adapt the grammar to LaTeX code
- [ ] Testcases have not been added yet

As such the code is not quite ready to be added yet. The generated parser was tested on

### CODE REVIEW

File: grammar/Modelica.g [hunk 1]
```diff
@@ -0,0 +1,485 @@
+// Define a grammar called Hello
```
Comment (adrpo) [inline grammar/Modelica.g]: I guess we don't need that comment.

File: grammar/build.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh -x
+
+export CLASSPATH=.:$CLASSPATH
+make Modelica.g4
+# Test as ANTLR3 grammar; checking k<=2
+java org.antlr.Tool -Xconversiontimeout 10000 -report Modelica.g || exit 1
+java -Xmx500M org.antlr.v4.Tool Modelica.g4 || exit 1
+javac -g *.java || exit 1
+# java -Xmx500M org.antlr.v4.gui.TestRig Modelica stored_definition -tokens test.mo || exit 1
+java -Xmx500M org.antlr.v4.gui.TestRig Modelica stored_definition $1 test.mo || exit 1
```
Comment (adrpo) [inline grammar/build.sh]: maybe we can run the parser on the MSL at least

Comment (sjoelund) [reply grammar/build.sh]: The Jenkins job runs it on MSL, Buildings and Synchronous

Comment (adrpo) [reply grammar/build.sh]: Ah, OK!

File: grammar/Modelica.g (hunk 1 above)
Comment (adrpo) [reply grammar/Modelica.g]: I will remove the comment myself.

**Comment by HansOlsson** [comment]
> There is no script to adapt the grammar to LaTeX code

As I see it that would give the clear advantage of having a tested grammar in the specification; with a single source of truth.
Including an external code-fragment is straightforward, but I understand that there are a number of differences that should be resolved:

- Different meta-symbols (we could change them if necessary)
- Underscore vs. dash in grammar productions (it was changed to dash for readability reasons).
- The grammar in the specification is split into sub-sections, which seems fairly simple - and I could not see that ANTLR has an include-directives allowing that split.

The last point made me realize that this is not only about the grammar-appendix, as there are a number of grammar-fragments spread over the specification (some are slightly modified - ignore them for now). The script could generate them as well - provided we allow it to generate "overlapping" fragments.
(Including fragments in LaTeX is also possible, but I thought it was primarily line-based which doesn't work well.) 

However, I think we should still keep the generated files for the specification in Git, and only verify them - since otherwise it will add another requirement.

**Comment by sjoelund** [comment]
It should have been a draft PR, but it's a new GitHub feature that messes with Jenkins :)

**Comment by sjoelund** [comment]
I think the most important thing is to decide if we can simplify the grammar by allowing `expression` as a stand-alone statement (instead of only function calls). That's the most tricky part in the grammar and with a flat Modelica / equation code representation it may be convenient to allow it.

Using ANTLR backtracking in the official grammar may look a bit weird.
As would having a semantic rule that checked it after the expression was parsed.
And duplicating the expression parser to have a special case for function calls as the expression is also really inconvenient.
Using ANTLRv4 style without declaring backtracking but it being used automatically would not make it an `LL(k)` grammar anymore, but rather an `ALL(*)` grammar.

**Comment by adrpo** [comment]
I will review it again, but you're right we should find some other reviewer that hasn't been involved in the development.

**Comment by sjoelund** [comment]
@henrikt-ma - this is the PR for the grammar. Some fixes from the existing grammar in the PDF, some changes.

**Comment by HansOlsson** [comment]
> @henrikt-ma - this is the PR for the grammar. Some fixes from the existing grammar in the PDF, some changes.

It's complicated when a PR mixes cleanup, new features, and other changes; and we need to at least be able to tell those apart.

**Comment by christoff-buerger** [comment]
> der_class_specifier was merged into short_class_specifier to get
k=2 instead of k=3.

Your proposed solution for this is **wrong**. The problem is that `short-class-specifier` are part of `short-class-definition` which in turn can be `element-redeclaration`s, but `der-class-specifier` are not permitted in such.

You either factor the common prefix `IDENT "="` correctly out -- i.e., at the level of `class-specifier` -- or -- what I would recommend to keep the level of unnecessary changes and potential for errors low -- just leave it LL(3) for now.

**Comment by HansOlsson** [comment]
The 

> extends_clause was modified from the 3.4 grammar (no rule annotation exists)

should be turned into a separate PR so that it can also be merged to the 3.4 maintenance branch.

The reason for the change is that the grammar part of the specification now uses syntax highlighting instead of manual highlighting, which requires separating annotation as a keyword and annotation as a grammar-production (the latter renamed to annotation-comment, but it was missed in one case).

**Comment by adrpo** [comment]
Really? a docker image without "make" is used?
@sjoelund: can you add that so we don't have to duplicate the commands in the Makefile again in .CI/Jenkinsfile

**Comment by HansOlsson** [comment]
> Really? a docker image without "make" is used?
> @sjoelund: can you add that so we don't have to duplicate the commands in the Makefile again in .CI/Jenkinsfile

This made me realize a minor issue.

Auto-generating the tex-grammar from the ANTLR file is good to ensure that it is correct.

A minor down-side is that if "done right" it requires that everyone installs more tools locally to generate pdf-files even if they don't touch the grammar. The simplest possibility to handle that is to also have the generated tex-files in git, even though it is against best practices. (Note: This only works if the generated files will have the same contents every time, i.e. no time-stamps at the start.)

**Comment by adrpo** [comment]
Is only sed that is needed to generate the .g.tex files, the ANTLR is only needed for testing.
But of course, we could split the listing we have now in BXX_name.tex and then generate those from .g.
Maybe as first iteration this would be good so we could actually see the diff of what we changed.

**Comment by adrpo** [comment]
I will do the changes i talked about so you can see what I mean.

**Comment by sjoelund** [comment]
@HansOlsson Yes, one possibility is to include auto-generated code in git. Another would be to by default skip the grammar appendix if the file is missing, adding a placeholder.

**Comment by adrpo** [comment]
The first diff looks like this (what it is now vs. generated files). I'll try to minimize it now.

``` diff
diff --git a/chapters/B1_lexical_conventions.tex b/chapters/B1_lexical_conventions.tex
index 88b982f..933e3f0 100644
--- a/chapters/B1_lexical_conventions.tex
+++ b/chapters/B1_lexical_conventions.tex
@@ -1,16 +1,51 @@
+% generated file! DO NOT EDIT! Edit the ../grammar/.g file instead
 \begin{lstlisting}[language=grammar]
-IDENT = NONDIGIT { DIGIT | NONDIGIT } | Q-IDENT
-Q-IDENT = "'" ( Q-CHAR | S-ESCAPE ) { Q-CHAR | S-ESCAPE | """ } "'"
-NONDIGIT = "_" | letters "a" to "z" | letters "A" to "Z"
-STRING = """ { S-CHAR | S-ESCAPE } """
-S-CHAR = see below
-Q-CHAR = NONDIGIT | DIGIT | "!" | "#" | "$" | "%" | "&" | "(" | ")" | "*" | "+" | "," |
-         "-" | "." | "/" | ":" | ";" | "<" | ">" | "=" | "?" | "@" | "[" | "]" | "^" |
-        "{" | "}" | "|" | "~" | " "
-S-ESCAPE = "\'" | "\"" | "\?" | "\\" |
-         "\a" | "\b" | "\f" | "\n" | "\r" | "\t" | "\v"
-DIGIT = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
-UNSIGNED-INTEGER = DIGIT { DIGIT }
-UNSIGNED-NUMBER = UNSIGNED-INTEGER [ "." [ UNSIGNED-INTEGER ] ]
-        [ ( "e" | "E" ) [ "+" | "-" ] UNSIGNED-INTEGER ]
+lexer grammar B1-lexical-conventions;
+
+// Whitespace and comments
+
+BOM : '\u00EF' '\u00BB' '\u00BF' ;
+
+WS : ( ' ' | '\t' | NL )+ { $channel=HIDDEN; }
+  ;
+
+LINE-COMMENT
+    : '//' ( ~('\r'|'\n')* ) (NL|EOF) { $channel=HIDDEN; }
+    ;
+
+ML-COMMENT
+    :   '/*' (options {greedy=false;} : .)* '*/' { $channel=HIDDEN;  }
+    ;
+
+fragment
+NL: '\r\n' | '\n' | '\r';
+
+// Lexical units except for keywords
+
+IDENT : NONDIGIT ( DIGIT | NONDIGIT )* | Q-IDENT ;
+
+fragment NONDIGIT : '_' | 'a' .. 'z' | 'A' .. 'Z' ;
+
+STRING : '"' ( S-CHAR | S-ESCAPE )* '"' ;
+
+fragment S-CHAR : NL | ~('\r' | '\n' | '\\' | '"'); // Unicode other than " and \
+
+fragment DIGIT : '0' .. '9' ;
+
+fragment Q-IDENT : '\'' ( Q-CHAR | S-ESCAPE ) ( Q-CHAR | S-ESCAPE | '"' )* '\'' ;
+
+fragment Q-CHAR
+   : NONDIGIT | DIGIT | '!' | '#' | '$' | '%' | '&' | '(' | ')' | '*'
+   | '+' | ',' | '-' | '.' | '/' | ':' | ';' | '<' | '>' | '=' | '?'
+   | '@' | '[' | ']' | '^' | '{' | '}' | '|' | '~' | ' '
+   ;
+fragment S-ESCAPE : '\\'
+  ( '\'' | '"' | '?' | '\\' | 'a' | 'b' | 'f' | 'n' | 'r' | 't' | 'v')
+  ;
+
+fragment UNSIGNED-INTEGER : DIGIT+ ;
+fragment EXPONENT : ( 'e' | 'E' ) ( '+' | '-' )? DIGIT+ ;
+
+UNSIGNED-NUMBER : DIGIT+ ( '.' (DIGIT)* )? ( EXPONENT )? ;
+
 \end{lstlisting}
diff --git a/chapters/B21_stored_definition.tex b/chapters/B21_stored_definition.tex
index 153f155..f5b6962 100644
--- a/chapters/B21_stored_definition.tex
+++ b/chapters/B21_stored_definition.tex
@@ -1,5 +1,14 @@
+% generated file! DO NOT EDIT! Edit the ../grammar/.g file instead
 \begin{lstlisting}[language=grammar]
-stored-definition:
-   [ within [ name ] ";" ]
-   { [ final ] class-definition ";" }
+parser grammar B21-stored-definition;
+
+// Stored Definition - Within
+
+stored-definition  :
+  '\uFEFF'? // BOM
+  ( 'within' name? ';' )?
+  ( 'final'? class-definition ';' )*
+  EOF
+  ;
+  
 \end{lstlisting}
diff --git a/chapters/B22_class_definition.tex b/chapters/B22_class_definition.tex
index 8577977..5f8cddd 100644
--- a/chapters/B22_class_definition.tex
+++ b/chapters/B22_class_definition.tex
@@ -1,70 +1,99 @@
+% generated file! DO NOT EDIT! Edit the ../grammar/.g file instead
 \begin{lstlisting}[language=grammar]
+parser grammar B22-class-definition;
+
+// Class Definition
+
+start-rule: class-definition;
+
 class-definition :
-   [ encapsulated ] class-prefixes
-   class-specifier
+  'encapsulated'? class-prefixes
+  class-specifier
+  ;
 
 class-prefixes :
-   [ partial ]
-   ( class | model | [ operator ] record | block | [ expandable ] connector | type |
-   package | [ pure | impure ] [ operator ] function | operator )
+  'partial'?
+  ( 'class' | 'model' | 'operator'? 'record' | 'block' | 'expandable'? 'connector' |
+  'type' | 'package' | ( 'pure' | 'impure' )? 'operator'? 'function' | 'operator' )
+  ;
 
 class-specifier :
-   long-class-specifier | short-class-specifier | der-class-specifier
+  long-class-specifier | short-class-specifier | der-class-specifier
+  ;
 
 long-class-specifier :
-   IDENT string-comment composition end IDENT
-   | extends IDENT [ class-modification ] string-comment composition
-   end IDENT
+  IDENT string-comment composition 'end' IDENT
+  | 'extends' IDENT class-modification? string-comment composition
+    'end' IDENT
+  ;
 
 short-class-specifier :
-   IDENT "=" base-prefix type-specifier [ array-subscripts ]
-   [ class-modification ] comment
-   | IDENT "=" enumeration "(" ( [ enum-list ] | ":" ) ")" comment
+  IDENT '='
+  ( 
+    base-prefix type-specifier array-subscripts? class-modification?
+  | 'enumeration' '(' ( enum-list? | ':' ) ')'
+  )
+  comment
+  ;
 
 der-class-specifier :
-   IDENT "=" der "(" type-specifier "," IDENT { "," IDENT } ")" comment
+  IDENT '=' 'der' '(' type-specifier ',' IDENT ( ',' IDENT )* ')' comment
+  ;
 
-base-prefix :
-   [ input | output ]
+base-prefix
+  : ( 'input' | 'output' )?
+  ;
 
-enum-list : enumeration-literal { "," enumeration-literal}
+enum-list :
+  enumeration-literal ( ',' enumeration-literal )*
+  ;
 
-enumeration-literal : IDENT comment
+enumeration-literal :
+  IDENT comment
+  ;
 
 composition :
-   element-list
-   { public element-list |
-     protected element-list |
-     equation-section |
-     algorithm-section
-   }
-   [ external [ language-specification ]
-   [ external-function-call ] [ annotation-comment ] ";" ]
-   [ annotation-comment ";" ]
+  element-list
+  ( 'public' element-list |
+    'protected' element-list |
+    equation-section |
+    algorithm-section
+  )*
+  ( 'external' language-specification
+  external-function-call? annotation-comment? ';' )?
+  ( annotation-comment ';' )?
+  ;
 
 language-specification :
-   STRING
+  STRING
+  ;
 
 external-function-call :
-   [ component-reference "=" ]
-   IDENT "(" [ expression-list ] ")"
+  ( component-reference '=' )?
+  IDENT '(' expression-list? ')'
+  ;
 
 element-list :
-   { element ";" }
+  ( element ';' )*
+  ;
 
 element :
-   import-clause |
-   extends-clause |
-   [ redeclare ]
-   [ final ]
-   [ inner ] [ outer ]
-   ( class-definition | component-clause |
-   replaceable ( class-definition | component-clause )
-   [ constraining-clause comment ] )
+  import-clause |
+  extends-clause |
+  'redeclare'?
+  'final'?
+  'inner'? 'outer'?
+  ( ( class-definition | component-clause) |
+  'replaceable' ( class-definition | component-clause )
+  (constraining-clause comment)? )
+  ;
 
 import-clause :
-   import ( IDENT "=" name | name ["." ( "*" | "{" import-list "}" ) ] ) comment
+  'import' ( IDENT '=' name | name ('.' ( '*' | '{' import-list '}' ) | '.*')? ) comment
+  ;
 
 import-list :
-   IDENT { "," IDENT }
+  IDENT ( ',' IDENT )*
+  ;
+
 \end{lstlisting}
diff --git a/chapters/B23_extends.tex b/chapters/B23_extends.tex
index f634d21..ad846b1 100644
--- a/chapters/B23_extends.tex
+++ b/chapters/B23_extends.tex
@@ -1,7 +1,17 @@
+% generated file! DO NOT EDIT! Edit the ../grammar/.g file instead
 \begin{lstlisting}[language=grammar]
+parser grammar B23-extends;
+
+// Extends
+
+start-rule: extends-clause;
+
 extends-clause :
-   extends type-specifier [ class-modification ] [ annotation-comment ]
+  'extends' type-specifier class-modification? annotation-comment?
+  ;
 
 constraining-clause :
-   constrainedby type-specifier [ class-modification ]
+  'constrainedby' type-specifier class-modification?
+  ;
+
 \end{lstlisting}
diff --git a/chapters/B24_component_clause.tex b/chapters/B24_component_clause.tex
index 888575f..6911095 100644
--- a/chapters/B24_component_clause.tex
+++ b/chapters/B24_component_clause.tex
@@ -1,19 +1,34 @@
+% generated file! DO NOT EDIT! Edit the ../grammar/.g file instead
 \begin{lstlisting}[language=grammar]
-component-clause:
-   type-prefix type-specifier [ array-subscripts ] component-list
+parser grammar B24-component-clause;
+
+// Component Clause
+
+start-rule: component-clause;
+
+component-clause :
+  type-prefix type-specifier array-subscripts? component-list
+  ;
 
 type-prefix :
-   [ flow | stream ]
-   [ discrete | parameter | constant ] [ input | output ]
+  ( 'flow' | 'stream' )?
+  ( 'discrete' | 'parameter' | 'constant' )? ( 'input' | 'output' )?
+  ;
+
 component-list :
-   component-declaration { "," component-declaration }
+  component-declaration ( ',' component-declaration )*
+  ;
 
 component-declaration :
-   declaration [ condition-attribute ] comment
+  declaration condition-attribute? comment
+  ;
 
-condition-attribute:
-   if expression
+condition-attribute :
+  'if' expression
+  ;
 
 declaration :
-   IDENT [ array-subscripts ] [ modification ]
+  IDENT array-subscripts? modification?
+  ;
+
 \end{lstlisting}
diff --git a/chapters/B25_modification.tex b/chapters/B25_modification.tex
index ae91f8e..4fc6e44 100644
--- a/chapters/B25_modification.tex
+++ b/chapters/B25_modification.tex
@@ -1,39 +1,58 @@
+% generated file! DO NOT EDIT! Edit the ../grammar/.g file instead
 \begin{lstlisting}[language=grammar]
+parser grammar B25-modification;
+
+// Modification
+
+start-rule: modification;
+
 modification :
-   class-modification [ "=" expression ]
-   | "=" expression
-   | ":=" expression
+  class-modification ( '=' expression )?
+  | '=' expression
+  | ':=' expression
+  ;
 
 class-modification :
-   "(" [ argument-list ] ")"
+  '(' argument-list? ')'
+  ;
 
-argument-list :
-   argument { "," argument }
+argument-list
+  : argument ( ',' argument )*
+  ;
 
-argument :
-   element-modification-or-replaceable
-   | element-redeclaration
+argument
+  : element-modification-or-replaceable
+  | element-redeclaration
+  ;
 
-element-modification-or-replaceable:
-   [ each ] [ final ] ( element-modification | element-replaceable )
+element-modification-or-replaceable
+  : 'each'? 'final'? ( element-modification | element-replaceable )
+  ;
 
-element-modification :
-   name [ modification ] string-comment
+element-modification
+  : name modification? string-comment
+  ;
 
 element-redeclaration :
-   redeclare [ each ] [ final ]
-   ( short-class-definition | component-clause1 | element-replaceable )
+  'redeclare' 'each'? 'final'?
+  ( ( short-class-definition | component-clause1 ) | element-replaceable )
+  ;
 
-element-replaceable:
-   replaceable ( short-class-definition | component-clause1 )
-   [ constraining-clause ]
+element-replaceable :
+  'replaceable' ( short-class-definition | component-clause1 )
+  constraining-clause?
+  ;
 
 component-clause1 :
-   type-prefix type-specifier component-declaration1
+  type-prefix type-specifier component-declaration1
+  ;
+
+component-declaration1
+  : declaration comment
+  ;
 
-component-declaration1 :
-   declaration comment
+short-class-definition
+  : class-prefixes short-class-specifier
+  ;
 
-short-class-definition :
-   class-prefixes short-class-specifier
-\end{lstlisting}
\ No newline at end of file
+\end{lstlisting}
diff --git a/chapters/B26_equations.tex b/chapters/B26_equations.tex
index 54f5ea9..8f18f16 100644
--- a/chapters/B26_equations.tex
+++ b/chapters/B26_equations.tex
@@ -1,87 +1,111 @@
+% generated file! DO NOT EDIT! Edit the ../grammar/.g file instead
 \begin{lstlisting}[language=grammar]
+parser grammar B26-equations;
+
+// Equations
+
+start-rule: equation-section;
+
 equation-section :
-   [ initial ] equation { equation ";" }
+  'initial'? 'equation' ( equation ';' )*
+  ;
 
 algorithm-section :
-   [ initial ] algorithm { statement ";" }
+  'initial'? 'algorithm' ( statement ';' )*
+  ;
 
 equation :
-   ( simple-expression "=" expression
-     | if-equation
-     | for-equation
-     | connect-clause
-     | when-equation
-     | component-reference function-call-args )
-   comment
+  (
+    simple-expression ( '=' expression )? // CHANGE: optional last part to support component-reference function-call-args via simple-expression
+  | if-equation
+  | for-equation
+  | connect-clause
+  | when-equation
+  ) comment
+  ;
 
 statement :
-   ( component-reference ( ":=" expression | function-call-args )
-     | "(" output-expression-list ")" ":=" component-reference function-call-args
-     | break
-     | return
-     | if-statement
-     | for-statement
-     | while-statement
-     | when-statement )
-   comment
+  ( component-reference ( ':=' expression | function-call-args )
+  | '(' output-expression-list ')' ':=' component-reference function-call-args
+  | 'break'
+  | 'return'
+  | if-statement
+  | for-statement
+  | while-statement
+  | when-statement )
+  comment
+  ;
 
 if-equation :
-   if expression then
-     { equation ";" }
-   { elseif expression then
-     { equation ";" }
-   }
-   [ else
-     { equation ";" }
-   ]
-   end if
+  'if' expression 'then'
+    ( equation ';' )*
+  ( 'elseif' expression 'then'
+    ( equation ';' )*
+  )*
+  ( 'else'
+    ( equation ';' )*
+  )?
+  'end' 'if'
+  ;
 
 if-statement :
-   if expression then
-     { statement ";" }
-   { elseif expression then
-     { statement ";" }
-   }
-   [ else
-     { statement ";" }
-   ]
-   end if
+  'if' expression 'then'
+    ( statement ';' )*
+  ( 'elseif' expression 'then'
+    ( statement ';' )*
+  )*
+  ( 'else'
+    ( statement ';' )*
+  )?
+  'end' 'if'
+  ;
 
 for-equation :
-   for for-indices loop
-     { equation ";" }
-   end for
+  'for' for-indices 'loop'
+    ( equation ';' )*
+  'end' 'for'
+  ;
 
 for-statement :
-   for for-indices loop
-     { statement ";" }
-   end for
+  'for' for-indices 'loop'
+    ( statement ';' )*
+  'end' 'for'
+  ;
 
 for-indices :
-   for-index {"," for-index}
+  for-index ( ',' for-index )*
+  ;
 
-for-index:
-   IDENT [ in expression ]
+for-index :
+  IDENT ( 'in' expression )?
+  ;
 
 while-statement :
-   while expression loop
-   { statement ";" }
-   end while
+  'while' expression 'loop'
+    ( statement ';' )*
+  'end' 'while'
+  ;
 
 when-equation :
-   when expression then
-     { equation ";" }
-   { elsewhen expression then
-     { equation ";" } }
-   end when
+  'when' expression 'then'
+    ( equation ';' )*
+  ( 'elsewhen' expression 'then'
+    ( equation ';' )*
+  )*
+  'end' 'when'
+  ;
 
 when-statement :
-   when expression then
-     { statement ";" }
-   { elsewhen expression then
-     { statement ";" } }
-   end when
+  'when' expression 'then'
+    ( statement ';' )*
+  ( 'elsewhen' expression 'then'
+    ( statement ';' )*
+  )*
+  'end' 'when'
+  ;
 
 connect-clause :
-   connect "(" component-reference "," component-reference ")"
+  'connect' '(' component-reference ',' component-reference ')'
+  ;
+
 \end{lstlisting}
diff --git a/chapters/B27_expressions.tex b/chapters/B27_expressions.tex
index 0386010..d37410d 100644
--- a/chapters/B27_expressions.tex
+++ b/chapters/B27_expressions.tex
@@ -1,105 +1,148 @@
+% generated file! DO NOT EDIT! Edit the ../grammar/.g file instead
 \begin{lstlisting}[language=grammar]
+parser grammar B27-expressions;
+
+// Expressions
+
+start-rule: expression;
+
 expression :
-   simple-expression
-   | if expression then expression { elseif expression then expression }
-   else expression
+  simple-expression
+  | 'if' expression 'then' expression ( 'elseif' expression 'then' expression )*
+    'else' expression
+  ;
 
 simple-expression :
-   logical-expression [ ":" logical-expression [ ":" logical-expression ] ]
+  logical-expression ( ':' logical-expression ( ':' logical-expression )? )?
+  ;
 
 logical-expression :
-   logical-term { or logical-term }
+  logical-term ( 'or' logical-term )*
+  ;
 
 logical-term :
-   logical-factor { and logical-factor }
+  logical-factor ( 'and' logical-factor )*
+  ;
 
 logical-factor :
-   [ not ] relation
+  'not'? relation
+  ;
 
 relation :
-   arithmetic-expression [ relational-operator arithmetic-expression ]
+  arithmetic-expression ( relational-operator arithmetic-expression )?
+  ;
 
 relational-operator :
-   "<" | "<=" | ">" | ">=" | "==" | "<>"
+  '<' | '<=' | '>' | '>=' | '==' | '<>'
+  ;
 
 arithmetic-expression :
-   [ add-operator ] term { add-operator term }
+  add-operator? term ( add-operator term )*
+  ;
 
 add-operator :
-   "+" | "-" | ".+" | ".-"
+  '+' | '-' | '.+' | '.-'
+  ;
 
 term :
-   factor { mul-operator factor }
+  factor ( mul-operator factor )*
+  ;
 
 mul-operator :
-   "*" | "/" | ".*" | "./"
+  '*' | '/' | '.*' | './'
+  ;
 
 factor :
-   primary [ ("^" | ".^") primary ]
+  primary ( ('^' | '.^') primary )?
+  ;
 
 primary :
-   UNSIGNED-NUMBER
-   | STRING
-   | false
-   | true
-   | ( component-reference | der | initial | pure ) function-call-args
-   | component-reference
-   | "(" output-expression-list ")"
-   | "[" expression-list { ";" expression-list } "]"
-   | "{" array-arguments "}"
-   | end
-
-type-specifier : ["."] name
-
-name : IDENT { "." IDENT }
+  UNSIGNED-NUMBER
+  | STRING
+  | 'false'
+  | 'true'
+  | ( 'der' | 'initial' | 'pure' ) function-call-args
+  | component-reference function-call-args?
+  | '(' output-expression-list ')'
+  | '[' expression-list ( ';' expression-list )* ']'
+  | '{' array-arguments '}'
+  | 'end'
+  ;
+
+type-specifier :
+  '.'? name
+  ;
+
+name :
+  IDENT ( '.' IDENT )*
+  ;
 
 component-reference :
-   [ "." ] IDENT [ array-subscripts ] { "." IDENT [ array-subscripts ] }
+  '.'? IDENT array-subscripts? ( '.' IDENT array-subscripts? )*
+  ;
 
 function-call-args :
-   "(" [ function-arguments ] ")"
+  '(' function-arguments? ')'
+  ;
 
-function-arguments :
-expression [ "," function-arguments-non-first | for for-indices ]
-   | function name "(" [ named-arguments ] ")" [ "," function-arguments-non-first ]
-   | named-arguments
+function-arguments
+  : expression ( ',' function-arguments-non-first | 'for' for-indices )?
+  | function-partial-application (',' function-arguments-non-first)?
+  | named-arguments
+  ;
 
-function-arguments-non-first :
-   function-argument [ "," function-arguments-non-first ]
-   | named-arguments
+function-arguments-non-first
+  : ( function-argument (',' function-arguments-non-first)?
+    | named-arguments
+    )
+  ;
 
 array-arguments :
-   expression [ "," array-arguments-non-first | for for-indices ]
+  expression ( ( ',' expression )* | 'for' for-indices ) // CHANGE: repetition instead of a new rule
+  ;
 
-array-arguments-non-first :
-   expression [ "," array-arguments-non-first ]
+named-arguments :
+  named-argument ( ',' named-argument )* // CHANGE: repetition
+  ;
 
-named-arguments: named-argument [ "," named-arguments ]
+named-argument :
+  IDENT '=' function-argument
+  ;
 
-named-argument: IDENT "=" function-argument
+function-argument
+  : function-partial-application
+  | expression
+  ;
 
-function-argument :
-   function name "(" [ named-arguments ] ")" | expression
+function-partial-application // CHANGE: This is a new rule; refactored out
+  : 'function' type-specifier '(' named-arguments? ')'
+  ;
 
-output-expression-list:
-   [ expression ] { "," [ expression ] }
+output-expression-list :
+  expression? ( ',' expression? )*
+  ;
 
 expression-list :
-   expression { "," expression }
+  expression ( ',' expression? )*
+  ;
 
 array-subscripts :
-   "[" subscript { "," subscript } "]"
+  '[' subscript ( ',' subscript )* ']'
+  ;
 
 subscript :
-   ":" | expression
+  ':' | expression
+  ;
 
 comment :
-   string-comment [ annotation-comment ]
+  string-comment annotation-comment?
+  ;
 
 string-comment :
-[ STRING { "+" STRING } ]
+  ( STRING ( '+' STRING )* )?
+  ;
 
 annotation-comment :
-   annotation class-modification
+  'annotation' class-modification
+  ;
 \end{lstlisting}
-

```

**Comment by adrpo** [comment]
@sjoelund: I know, but for my sanity and debugging I will do each phase separately :)

**Comment by adrpo** [comment]
The parser diffs are rather small now.

**Comment by henrikt-ma** [comment]
> > @henrikt-ma - this is the PR for the grammar. Some fixes from the existing grammar in the PDF, some changes.
> 
> It's complicated when a PR mixes cleanup, new features, and other changes; and we need to at least be able to tell those apart.

For the MCP work on [Flat Modelica](https://github.com/modelica/ModelicaSpecification/tree/MCP/0031/RationaleMCP/0031) and eFMI, we just wanted something that would roughly correspond to current Modelica as a starting point.  Details are not important at this point, so the assumption is that anything found in this PR will do.

**Comment by HansOlsson** [comment]
What is the plan for this one?
It should be converted to a draft if it is still will be worked on, and otherwise closed.

**Comment by HansOlsson** [comment]
> What is the plan for this one? It should be converted to a draft if it is still will be worked on, and otherwise closed.

Closing, No response after a year.

**Pull request closed without merging**